### ISSUE TITLE
pgtype zero value

### ISSUE THREAD

**Issue by aravindc26** [jackc/pgx#296]
I wonder why there is a need for `Undefined` status in pgtype ? I believe `Null` would have been more appropriate one. Most unintialized values in the application that I am working on, can be defaulted to null;   Can you also please explain the need for an extra status `Undefined` ?

**Comment by hsyed** [comment]
Undefined is needed for code generation and generic models. In a generic model the set of values are placeholders. Undefined means that the placeholder wasn't initialised -- it has no value. Null is a value.  

Think about writing a generic update method, you would be stuck with null / zero -- how do you ignore certain fields ? .

**Comment by jackc** [comment]
Exactly what @hsyed said above.

**Comment by ronaldslc** [comment]
In V2, We used nil pointer to represent uninitialised values, I wonder what is wrong or not as good with that approach. 

With the new pgtype approach, if we have structures where uninitialised values are potentially unused in a lot of cases, it would now take up extra memory as go initialises them with defaults.

**Comment by jackc** [comment]
I don't think there is anything _wrong_ with using nil pointers, but I suspect if you were to profile code written both ways you would find that the pgtype style more efficient for most cases.

Here's an example in pgx v2 style:

```go
struct Widget {
	// other fields
	Num *pgx.NullInt64
}
```

When `Num` is undefined it costs 8 bytes (assuming 64-bit architecture) as part of the `Widget` struct and requires no additional allocations. When it is null or present it still costs the 8-bytes pointer as part of `Widget`, but it also costs an additional allocation of 9 bytes (8 for Int, 1 for Valid). Due to memory alignment concerns it probably will allocate 16-bytes. Accessing the field will require a pointer dereference.

And here's the example in v3 style:

```go
struct Widget {
	// other fields
	Num pgtype.Int8
}
```

The pgtype.Int8 struct is 9 bytes in size regardless of undefined, null, or present status. Memory alignment will again probably cause it to really cost 16 bytes. It never occurs an additional allocation and it does not require a pointer dereference to access.

So to summarize, v2 style saves 8 bytes when the field is undefined/nil. But is any other case it costs an additional 8 bytes, an additional memory allocation, and a pointer dereference on each access (it also is less cache-friendly due to less data locality).
